File: mobile/ios/OTravelz/Features/Notifications/NotificationModels.swift
import Foundation

/// Request model for scheduling a local transit departure reminder.
public struct TransitReminderRequest: Sendable, Equatable {
    public let routeId: String
    public let routeNumber: String
    public let origin: String
    public let departureTime: String // HH:mm in IST
    public let offsetMinutes: Int

    public init(
        routeId: String,
        routeNumber: String,
        origin: String,
        departureTime: String,
        offsetMinutes: Int = 15
    ) {
        self.routeId = routeId
        self.routeNumber = routeNumber
        self.origin = origin
        self.departureTime = departureTime
        self.offsetMinutes = offsetMinutes
    }

    public var notificationId: String {
        let cleanTime = departureTime.replacingOccurrences(of: ":", with: "")
        return "rem_tr_\(routeId)_\(cleanTime)_\(offsetMinutes)m"
    }

    public var title: String {
        return "Route \(routeNumber) — Scheduled departure in \(offsetMinutes) min"
    }

    public var body: String {
        return "\(departureTime) IST from \(origin). Scheduled timetable only; check operator before travel."
    }

    public var deepLinkUri: String {
        return "otravelz://route/\(routeId)"
    }
}

/// Result of a reminder scheduling attempt.
public enum ReminderScheduleResult: Sendable, Equatable {
    case success(notificationId: String, scheduledEpochMs: Int64, message: String)
    case passedDeparture
    case permissionDenied
    case error(String)
}

/// Active reminder metadata for in-memory / UI tracking.
public struct ActiveReminderInfo: Sendable, Codable, Equatable {
    public let id: String
    public let routeId: String
    public let departureTime: String
    public let offsetMinutes: Int
    public let scheduledAtEpochMs: Int64

    public init(
        id: String,
        routeId: String,
        departureTime: String,
        offsetMinutes: Int,
        scheduledAtEpochMs: Int64
    ) {
        self.id = id
        self.routeId = routeId
        self.departureTime = departureTime
        self.offsetMinutes = offsetMinutes
        self.scheduledAtEpochMs = scheduledAtEpochMs
    }
}

File: mobile/android/src/main/kotlin/com/otravelz/android/notifications/NotificationModels.kt
package com.otravelz.android.notifications

import java.time.LocalDate

/**
 * Representation of a user-requested scheduled transit departure reminder.
 */
data class TransitReminderRequest(
    val routeId: String,
    val routeNumber: String,
    val origin: String,
    val departureTime: String, // HH:mm in IST
    val offsetMinutes: Int = 15,
    val targetDate: LocalDate = LocalDate.now(TransitReminderCalculator.ZONE_IST)
) {
    val notificationId: String
        get() = "rem_tr_${routeId}_${departureTime.replace(":", "")}_${offsetMinutes}m"

    val title: String
        get() = "Route $routeNumber — Scheduled departure in $offsetMinutes min"

    val body: String
        get() = "$departureTime IST from $origin. Scheduled timetable only; check operator before travel."

    val deepLinkUri: String
        get() = "otravelz://route/$routeId"
}

/**
 * Result of attempting to schedule a reminder.
 */
sealed interface ReminderScheduleResult {
    data class Success(
        val notificationId: String,
        val scheduledEpochMs: Long,
        val triggerMessage: String
    ) : ReminderScheduleResult

    data object PassedDeparture : ReminderScheduleResult
    data object PermissionDenied : ReminderScheduleResult
    data class Error(val message: String) : ReminderScheduleResult
}

/**
 * Lightweight representation of an actively scheduled reminder for UI presentation.
 */
data class ActiveReminderInfo(
    val id: String,
    val routeId: String,
    val departureTime: String,
    val offsetMinutes: Int,
    val scheduledAtEpochMs: Long
)
